### PR TITLE
PUB-3231 - Add MI Report Functionality in System Admin Dashboard

### PR DESCRIPTION
**Pull request by NatashaAlker**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PUB-3231

### Change description ###

Backend support for adding a new tile to the System Admin Dashboard that enables System Administrators to download MI Reports.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by alexbottenberg — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/publication/PublicationReportingController.java [hunk 1]
```diff
@@ -48,9 +49,10 @@ public PublicationReportingController(PublicationReportingService publicationRep
     @ApiResponse(responseCode = UNAUTHORISED_CODE, description = UNAUTHORISED_MESSAGE)
     @ApiResponse(responseCode = FORBIDDEN_CODE, description = FORBIDDEN_MESSAGE)
     @Operation(summary = "Returns MI data for artefacts")
-    @GetMapping("/mi-data")
-    public ResponseEntity<List<PublicationMiData>> getMiData() {
-        return ResponseEntity.ok().body(publicationReportingService.getMiData());
+    @GetMapping("/mi-data/{days}")
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/controllers/publication/PublicationReportingController.java]: 'days' should not be a path param. Path param is normally used for parameter to identify a resource like ID or name. 'days' is used for filtering so we should use query param. Also it is better to make the query param optional.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/service/publication/PublicationReportingService.java [hunk 2]
```diff
@@ -42,9 +42,9 @@ public PublicationReportingService(ArtefactRepository artefactRepository,
      * Retrieve artefact data for MI reporting.
      * @return MI artefact data as a list of PublicationMiData objects
      */
-    public List<PublicationMiData> getMiData() {
-        LocalDateTime publicationReceivedDate = LocalDate.now()
-            .minusDays(7)
+    public List<PublicationMiData> getMiData(int days) {
+        LocalDateTime publicationReceivedDate = days == -1 ? null : LocalDate.now()
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/service/publication/PublicationReportingService.java]: This logic here is using a bespoke magic number '-1' to handle the special case. Magic number is a bad practice in programming and not clear to the readers.

It is better to make 'days' in the controller and here a java Integer object (not int primitive type). This is done so 'days' can be null when not supplied so avoid the use of magic number.

All the unit, integration and functional tests also use '-1' and they need to be changed.

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java [hunk 3]
```diff
@@ -120,7 +120,7 @@ List<Artefact> findArtefactsByDisplayFrom(@Param(CURRENT_DATE_PARAM) LocalDate t
         + "provenance, sensitivity, sourceArtefactId, "
         + "supersededCount, type, contentDate, locationId, listType) "
         + "FROM Artefact "
-        + "WHERE lastReceivedDate >= :publicationReceivedDate")
+        + "WHERE CAST(:publicationReceivedDate AS timestamp) IS NULL OR lastReceivedDate >= :publicationReceivedDate")
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java]: We need to add a new test scenario in ArtefactRepositoryTest integration test to test this repository method when  publicationReceivedDate is null

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/service/publication/PublicationReportingServiceTest.java [hunk 4]
```diff
@@ -78,7 +78,7 @@ void testGetMiData()  {
         when(locationRepository.findAll()).thenReturn(List.of(location));
         when(artefactRepository.getMiData(any())).thenReturn(List.of(publicationMiData, publicationMiData2));
 
-        List<PublicationMiData> publicationMiDataList = publicationReportingService.getMiData();
+        List<PublicationMiData> publicationMiDataList = publicationReportingService.getMiData(-1);
```
Comment (KianKwa) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/service/publication/PublicationReportingServiceTest.java]: A few tests in this test class still using -1. Change it to null

File: src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java (hunk 3 above)
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/data/management/database/ArtefactRepository.java]: The comment above is still not addressed

File: src/test/java/uk/gov/hmcts/reform/pip/data/management/controllers/publication/PublicationReportingControllerTest.java [hunk 5]
```diff
@@ -46,9 +46,9 @@ void testMiDataReturnsSuccessfully() {
             Sensitivity.PUBLIC, UUID.randomUUID().toString(), 1, ArtefactType.GENERAL_PUBLICATION,
             LocalDateTime.now(), "NoMatch2", ListType.CIVIL_DAILY_CAUSE_LIST);
 
-        when(publicationReportingService.getMiData()).thenReturn(List.of(publicationMiData, publicationMiData2));
+        when(publicationReportingService.getMiData(-1)).thenReturn(List.of(publicationMiData, publicationMiData2));
```
Comment (KianKwa) [inline src/test/java/uk/gov/hmcts/reform/pip/data/management/controllers/publication/PublicationReportingControllerTest.java]: Avoid using -1 in the test parameter. Use null instead. Same issue in line 51

**Review by KianKwa — APPROVED** [review]
